src/OpenTelemetry.Extensions.Hosting/Implementation/DeferredServiceCollection.cs

```diff
@@ -0,0 +1,198 @@
+                    && ConvertToOtel(msRecord.Lifetime) == item.Lifetime)
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        public void Insert(int index, DeferredServiceDescriptor item)
+        {
+            this.Services.Insert(index, ConvertToMs(item));
+        }
+
+        public void RemoveAt(int index)
+        {
+            this.Services.RemoveAt(index);
+        }
+
+        public void Add(DeferredServiceDescriptor item)
+        {
+            this.Services.Add(ConvertToMs(item));
+        }
+
+        public void Clear()
+        {
+            this.Services.Clear();
+        }
+
+        public bool Contains(DeferredServiceDescriptor item)
+        {
+            return this.IndexOf(item) >= 0;
+        }
+
+        public void CopyTo(DeferredServiceDescriptor[] array, int arrayIndex)
+        {
+            if (array == null)
+            {
+                throw new ArgumentNullException(nameof(array));
+            }
+
+            var msRecords = new ServiceDescriptor[array.Length];
+
+            this.Services.CopyTo(msRecords, arrayIndex);
+
+            for (int i = arrayIndex; i < array.Length; i++)
+            {
+                array[i] = ConvertToOtel(msRecords[i]);
+            }
+        }
+
+        public bool Remove(DeferredServiceDescriptor item)
+        {
+            int index = this.IndexOf(item);
+
+            if (index >= 0)
+            {
+                this.Services.RemoveAt(index);
+                return true;
+            }
+
+            return false;
+        }
+
+        public IEnumerator<DeferredServiceDescriptor> GetEnumerator()
+        {
+            foreach (ServiceDescriptor serviceDescriptor in this.Services)
+            {
+                yield return ConvertToOtel(serviceDescriptor);
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+
+        private static DeferredServiceDescriptor ConvertToOtel(ServiceDescriptor serviceDescriptor)
+        {
+            if (serviceDescriptor == null)
+            {
+                return null;
+            }
+
+            if (serviceDescriptor.ImplementationInstance != null)
+            {
+                return new DeferredServiceDescriptor(serviceDescriptor.ServiceType, serviceDescriptor.ImplementationInstance);
+            }
+
+            var lifetime = ConvertToOtel(serviceDescriptor.Lifetime);
+
+            if (serviceDescriptor.ImplementationFactory != null)
+            {
+                return new DeferredServiceDescriptor(serviceDescriptor.ServiceType, serviceDescriptor.ImplementationFactory, lifetime);
+            }
+
+            return new DeferredServiceDescriptor(serviceDescriptor.ServiceType, serviceDescriptor.ImplementationType, lifetime);
+        }
+
+        private static DeferredServiceLifetime ConvertToOtel(ServiceLifetime serviceLifetime)
```

Comment:
Rename the methods to `ConvertToDeferredServiceLifetime` and `ConvertToDIServiceLifetime`?